sky/skylet/providers/fluidstack/node_provider.py

```diff
@@ -0,0 +1,147 @@
+import logging
+from threading import RLock
+from sky import authentication as auth
+from sky import authentication as auth
+import os
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_CLUSTER_NAME
+
+from sky.skylet.providers.fluidstack import fluidstack_utils
+
+import time
+
+logger = logging.getLogger(__name__)
+
+
+def synchronized(f):
+
+    def wrapper(self, *args, **kwargs):
+        self.lock.acquire()
+        try:
+            return f(self, *args, **kwargs)
+        finally:
+            self.lock.release()
+
+    return wrapper
+
+
+class FluidstackNodeProvider(NodeProvider):
+    """Node Provider for FluffyCloud."""
+
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.lock = RLock()
+        self.cached_nodes = {}
+        self.fluidstack_client = fluidstack_utils.FluidstackClient()
+        # Load credentials
+        public_key_path = os.path.expanduser(auth.PUBLIC_SSH_KEY_PATH)
+        self.ssh_public_key = None
+        with open(public_key_path, 'r') as f:
+            self.ssh_public_key = f.read().strip()
+        assert self.ssh_public_key is not None, 'Failed to load public key.'
+
+    def _get_filtered_nodes(self, tag_filters):
+        running_instances = self.fluidstack_client.list_instances()
+        self.cached_nodes = {}
+        if not tag_filters:
+            for instance in running_instances:
+                self.cached_nodes[instance['id']] = instance
+            return self.cached_nodes
+
+        for instance in running_instances:
+            for key, value in tag_filters.items():
+                tag = instance['tags'].get(key, None)
+                if tag and tag == value:
+                    self.cached_nodes[instance['id']] = instance
```

Comment:
Seems we have this tag filtering code in many places. Do we want to add it to the API of `fluidstack_client.list_instances`, i.e. `list_instances(tag_filters: Optional[Dict[str, str]])`